Tella/Scenes/Uwazi/ViewModel/UwaziViewModel.swift

```diff
@@ -0,0 +1,179 @@
+            }
+        }
+    }
+
+    /// To determine if the templates are already download or not reflect on the UI for template download list
+    /// - Parameter templates: Collection of CollectedTemplate to determine if it downloaded or not
+    func handleTemplateDownload(templates: inout [CollectedTemplate]) throws {
+        try templates.forEach { template in
+            let savedTemplateid = try self.getAllDownloadedTemplate()?.compactMap({$0.templateId})
+            if let savedTemplate = savedTemplateid,let templateId = template.templateId {
+                if savedTemplate.contains(templateId) {
+                    template.isDownloaded = 1
+                }
+            }
+        }
+    }
+    /// Save the template to the database
+    /// - Parameter template: The template that we need to save into the database
+    func saveTemplate( template: inout CollectedTemplate) {
+        do {
+            let savedTemplateid = try self.getAllDownloadedTemplate()?.compactMap({$0.templateId})
+            if let savedTemplate = savedTemplateid,let templateId = template.templateId {
+                // To only save the template if it is not already saved Not necessary because the UI will not have a download button if it is already downloaded
+                if !savedTemplate.contains(templateId) {
+                    let savedItem = try self.mainAppModel.vaultManager.tellaData.addUwaziTemplateWith(template: template)
+                    template = savedItem
+                }
+            }
+        } catch let error {
+            dump(error)
+        }
+    }
+    /// Delete the saved template from database using the template id of the template and changing the status of isDownloaded property to 0  for template listing view
+    /// - Parameter template: The CollectedTemplate Object and changing the status of isDownloaded property to 0
+    func deleteTemplate(template: inout CollectedTemplate) {
+        do {
+            if let templateId = template.templateId {
+                _ = try self.mainAppModel.vaultManager.tellaData.deleteAllUwaziTemplateWith(templateId: templateId)
+                template.isDownloaded = 0
+            }
+        } catch {
+
+        }
+    }
+    /// Get all the downloaded templates
+    /// - Returns: Collection of CollectedTemplate object which are stored in the database
+    func getAllDownloadedTemplate() throws -> [CollectedTemplate]? {
+        return try self.mainAppModel.vaultManager.tellaData.getAllUwaziTemplate()
+    }
+
+    /// Delete the saved template from database using the template id of the template for downloaded template listing view
+    /// - Parameter template: The template object which we need to delete
+    func deleteDownloadedTemplate(templateId: Int) {
+        do {
+            _ = try self.mainAppModel.vaultManager.tellaData.deleteAllUwaziTemplateWith(templateNo: templateId)
+            downloadedTemplates = downloadedTemplates.filter({$0.id != templateId})
```

Comment:
Here we could remove it by doing `downloadedTemplates.removeAll { $0.id == templateId }` instead of filtering and reassigning it, It is more direct that way.